src/library/scala/Array.scala

```diff
@@ -106,7 +106,9 @@ object Array {
    */
   def copy(src: AnyRef, srcPos: Int, dest: AnyRef, destPos: Int, length: Int): Unit = {
     val srcClass = src.getClass
-    if (srcClass.isArray && dest.getClass.isAssignableFrom(srcClass))
+    val destClass = dest.getClass
+    if (srcClass.isArray && (destClass.isAssignableFrom(srcClass) ||
```

Comment:
```suggestion
    if (srcClass.isArray && ((destClass eq srcClass) ||
```
When one of the two arrays is a primitive array, the only way one can be assignable from the other is if they are the same array type. 

Using `eq` here will therefore act both as a fast path for all typical copies, and handling the primitive cases. Then your new condition can deal with all the reference array types that don't make it into the fast path.